### PR TITLE
#2 Module 3 - Use mocked data from a JSON

### PR DESCRIPTION
**Pull request by agustingodoyc**
The objective is to escalate the MovieDB App to show all the information about the movies from the [API](https://developers.themoviedb.org/3/getting-started/introduction).

Screenshot_1696604597
Screenshot_1696604602

### CODE REVIEW

File: assets/mock_data/movies.json [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        "overview": "After a disease kills 98% of America's children, the surviving 2% develop superpowers and are placed in internment camps. A 16-year-old girl escapes her camp and joins a group of other teens on the run from the government.",
+        "popularity": 2158.368,
+        "poster_path": "/94RaS52zmsqaiAe1TG20pdbJCZr.jpg",
+        "release_date": "2018-07-25",
+        "title": "The Darkest Minds",
+        "video": false,
+        "vote_average": 7,
+        "vote_count": 2454
+    }
+]
```
Comment (Kevin-Guillen) [inline assets/mock_data/movies.json]: missing eof

File: assets/mock_data/genres.json [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        {
+            "id": 10752,
+            "name": "War"
+        },
+        {
+            "id": 37,
+            "name": "Western"
+        }
+    ]
+}
```
Comment (Kevin-Guillen) [inline assets/mock_data/genres.json]: missing eof

File: lib/repositories/implementation/genres_repository.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+import '../../data_models/genre.dart';
+import '../interface/i_genres_repository.dart';
+
+class GenresRepository implements IGenresRepository {
+  static const genres = 'assets/mock_data/genres.json';
+  List<Genre> genresList = [];
+
+  @override
+  Future<List<Genre>> fetchGenres() async {
+    if (genresList.isEmpty) {
```
Comment (Kevin-Guillen) [inline lib/repositories/implementation/genres_repository.dart:15]: we could use ternary operator here

File: lib/repositories/implementation/genres_repository.dart [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+      genresList = await parseGenres();
+    }
+    return genresList;
```
Comment (Kevin-Guillen) [inline lib/repositories/implementation/genres_repository.dart]: you can use ternary operator here

File: lib/routes/movie_details.dart [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+
+  @override
+  void initState() {
+    super.initState();
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      final MoviePreview movie =
+          ModalRoute.of(context)!.settings.arguments as MoviePreview;
+
+      final posterProvider = NetworkImage(movie.movie.posterUrl);
+      _updateImage(posterProvider);
```
Comment (Kevin-Guillen) [inline lib/routes/movie_details.dart]: you can move this into a new method to improve readability

File: lib/main.dart [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import 'package:flutter/material.dart';
-import 'package:the_movie_db_flutter/pages/movie_details.dart';
+import 'package:the_movie_db_flutter/routes/movie_details.dart';
+import 'package:the_movie_db_flutter/routes/movie_list.dart';
```
Comment (Kevin-Guillen) [inline lib/main.dart]: relative imports

File: lib/repositories/implementation/movies_repository.dart [hunk 7]
```diff
@@ -0,0 +1,27 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:flutter/services.dart';
+import 'package:the_movie_db_flutter/data_models/movie.dart';
+import 'package:the_movie_db_flutter/repositories/interface/i_movies_repository.dart';
```
Comment (Kevin-Guillen) [inline lib/repositories/implementation/movies_repository.dart]: relative imports

File: lib/routes/movie_details.dart [hunk 8]
```diff
@@ -0,0 +1,75 @@
+import 'package:flutter/material.dart';
+import 'package:the_movie_db_flutter/ui_models/movie_preview.dart';
```
Comment (Kevin-Guillen) [inline lib/routes/movie_details.dart]: relative import

File: lib/routes/movie_list.dart [hunk 9]
```diff
@@ -0,0 +1,85 @@
+import 'package:flutter/material.dart';
+import 'package:the_movie_db_flutter/repositories/implementation/genres_repository.dart';
+import 'package:the_movie_db_flutter/repositories/implementation/movies_repository.dart';
+import 'package:the_movie_db_flutter/repositories/interface/i_genres_repository.dart';
+import 'package:the_movie_db_flutter/repositories/interface/i_movies_repository.dart';
+import 'package:the_movie_db_flutter/widgets/movie_card.dart';
```
Comment (Kevin-Guillen) [inline lib/routes/movie_list.dart]: relative imports

File: lib/ui_models/movie_preview.dart [hunk 10]
```diff
@@ -0,0 +1,13 @@
+import 'package:the_movie_db_flutter/data_models/movie.dart';
```
Comment (Kevin-Guillen) [inline lib/ui_models/movie_preview.dart]: relative imports

File: lib/widgets/movie_card.dart [hunk 11]
```diff
@@ -0,0 +1,126 @@
+import 'dart:core';
+
+import 'package:flutter/material.dart';
+import 'package:the_movie_db_flutter/utils/ui_constants.dart';
+import 'package:the_movie_db_flutter/routes/movie_details.dart';
```
Comment (Kevin-Guillen) [inline lib/widgets/movie_card.dart]: relative imports

File: lib/widgets/movie_card.dart [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+                    child: Column(
+                      crossAxisAlignment: CrossAxisAlignment.start,
+                      children: [
+                        Padding(
+                          padding: const EdgeInsets.only(
+                            top: UIConstants.textsLargePadding,
+                          ),
+                          child: Text(
+                            widget.payload.movie.title,
+                            style: TextStyle(
```
Comment (Kevin-Guillen) [inline lib/widgets/movie_card.dart:88]: const

File: lib/widgets/movie_card.dart [hunk 13]
```diff
@@ -0,0 +1,126 @@
...
+                              color: currentColorScheme.onPrimary,
+                              fontWeight: FontWeight.bold,
+                            ),
+                          ),
+                        ),
+                        Text(
+                          widget.payload.movie.releaseDate,
+                          style: TextStyle(
```
Comment (Kevin-Guillen) [inline lib/widgets/movie_card.dart:96]: const

File: lib/widgets/movie_card.dart [hunk 14]
```diff
@@ -0,0 +1,126 @@
...
+                                .withOpacity(MovieCard.lessOpacity),
+                          ),
+                        ),
+                        Padding(
+                          padding: const EdgeInsets.symmetric(
+                            vertical: UIConstants.textsShortPadding,
+                          ),
+                          child: Text(
+                            widget.payload.movie.overview,
+                            style: TextStyle(
```
Comment (Kevin-Guillen) [inline lib/widgets/movie_card.dart:107]: const

File: lib/repositories/interface/i_movies_repository.dart [hunk 15]
```diff
@@ -0,0 +1,7 @@
+import '../../data_models/movie.dart';
+
+abstract class IMoviesRepository {
+  Future<List<Movie>> fetchMovies();
+
+  Future<List<Movie>> parseMovies();
```
Comment (matiasilveiro-gl) [inline lib/repositories/interface/i_movies_repository.dart]: I wouldn't add this `parseMovies` method as an interface method. You only need a `fetch` method, the implementation will take care of the **where** you get the data.
Same for the genres repo

File: lib/routes/movie_details.dart [hunk 16]
```diff
@@ -0,0 +1,75 @@
...
+            children: [
+              MovieHeader(
+                backdropUrl: payload.movie.backdropUrl,
+                posterUrl: payload.movie.posterUrl,
+              ),
+              MovieTitle(title: payload.movie.title),
+              VoteAverage(voteAverage: payload.movie.voteAverage),
+              FocusBox(
+                releaseDate: payload.movie.releaseDate,
+                genres: payload.genres.map((genre) => genre.name).join(genresSeparator),
```
Comment (matiasilveiro-gl) [inline lib/routes/movie_details.dart]: I'd move this to a getter method in this class, to improve readability

File: lib/routes/movie_details.dart [hunk 17]
```diff
@@ -0,0 +1,75 @@
...
+
+import '../widgets/focus_box.dart';
+import '../widgets/like_button.dart';
+import '../widgets/movie_header.dart';
+import '../widgets/movie_overview.dart';
+import '../widgets/movie_title.dart';
+import '../widgets/vote_average.dart';
+
+class MovieDetails extends StatefulWidget {
+  static const String routeName = "/movie";
```
Comment (FranciscoGaraventa) [inline lib/routes/movie_details.dart]: Move the route name to a new consta file that only managed routes like RoutesNames

File: lib/routes/movie_details.dart [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+      final posterProvider = NetworkImage(movie.movie.posterUrl);
+      _updateImage(posterProvider);
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final MoviePreview payload =
+        ModalRoute.of(context)!.settings.arguments as MoviePreview;
```
Comment (FranciscoGaraventa) [inline lib/routes/movie_details.dart]: why is call payload? this basically is a movie, you can get the movie outside build method

File: lib/ui_models/movie_preview.dart [hunk 19]
```diff
@@ -0,0 +1,13 @@
...
+
+import '../data_models/genre.dart';
+
+class MoviePreview {
```
Comment (FranciscoGaraventa) [inline lib/ui_models/movie_preview.dart]: You really need to have this MoviePreview class?

File: lib/utils/error_strings.dart [hunk 20]
```diff
@@ -0,0 +1,5 @@
+class ErrorStrings {
+  static const String errorMessage = 'Error';
+  static const String emptyMessage = 'Empty data';
+  static const String otherMessage = 'other';
+}
```
Comment (FranciscoGaraventa) [inline lib/utils/error_strings.dart]: EOF

File: lib/utils/route_constants.dart [hunk 21]
```diff
@@ -0,0 +1,4 @@
+class RouteConstants {
+  static const String movieRoute = "/movie";
+  static const String listRoute = '/';
+}
```
Comment (FranciscoGaraventa) [inline lib/utils/route_constants.dart]: EOF

**Review by FranciscoGaraventa — APPROVED** [review]
Approve with minor comment

**Review by matiasilveiro-gl — APPROVED** [review]
LGTM!

**Review by CIFUM007 — APPROVED** [review]

**Review by MarcosCGlobant — APPROVED** [review]

**Review by Kevin-Guillen — APPROVED** [review]

**Pull request merged**